### ISSUE TITLE
Segfault

### ISSUE THREAD

**Issue by tjingrant** [IntelLabs/FP8-Emulation-Toolkit#1]
Hi, I ran into the following segfault issue on a 3090 GPU with Ryzen 5950X CPU.

I ran the example script [here](https://github.com/IntelLabs/FP8-Emulation-Toolkit/blob/main/examples/inference/classifier/imagenet_test.py). The command I used is:
```
 python imagenet_test.py --data-path=/mnt/d/ILSVRC2012
```

Here's where the script segfaults (eval was fine, calibration seems to be the problem):

```Thread 80 "python" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff32824700 (LWP 27608)]
0x00007fffb298484b in void c10::function_ref<void (char**, long const*, long, long)>::callback_fn<at::native::AVX2::VectorizedLoop2d<at::native::AVX2::direct_copy_kernel(at::TensorIteratorBase&)::{lambda()#3}::operator()() const::{lambda()#14}::operator()() const::{lambda(float)#1}, at::native::AVX2::direct_copy_kernel(at::TensorIteratorBase&)::{lambda()#3}::operator()() const::{lambda()#3}::operator()() const::{lambda(at::vec::AVX2::Vectorized<float>)#2}> >(long, char**, long const*, long, long) () from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so```


Here's the full backtrace:


```#0  0x00007fffb298484b in void c10::function_ref<void (char**, long const*, long, long)>::callback_fn<at::native::AVX2::VectorizedLoop2d<at::native::AVX2::direct_copy_kernel(at::TensorIteratorBase&)::{lambda()#3}::operator()() const::{lambda()#14}::operator()() const::{lambda(float)#1}, at::native::AVX2::direct_copy_kernel(at::TensorIteratorBase&)::{lambda()#3}::operator()() const::{lambda()#3}::operator()() const::{lambda(at::vec::AVX2::Vectorized<float>)#2}> >(long, char**, long const*, long, long) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#1  0x00007fffae8e941d in at::TensorIteratorBase::serial_for_each(c10::function_ref<void (char**, long const*, long, long)>, at::Range) const () from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#2  0x00007fffae8e9957 in at::TensorIteratorBase::for_each(c10::function_ref<void (char**, long const*, long, long)>, long) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#3  0x00007fffb29d9826 in at::native::AVX2::direct_copy_kernel(at::TensorIteratorBase&) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#4  0x00007fffb29e65d0 in at::native::AVX2::copy_kernel(at::TensorIterator&, bool) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#5  0x00007fffaec26ff9 in at::native::copy_impl(at::Tensor&, at::Tensor const&, bool) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#6  0x00007fffaec28472 in at::native::copy_(at::Tensor&, at::Tensor const&, bool) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#7  0x00007fffaf7bf43f in at::_ops::copy_::call(at::Tensor&, at::Tensor const&, bool) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#8  0x00007fffc8eada69 in at::native::_pin_memory_cuda(at::Tensor const&, c10::optional<c10::Device>) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cuda_cpp.so
#9  0x00007fff836c3bed in at::(anonymous namespace)::(anonymous namespace)::wrapper___pin_memory(at::Tensor const&, c10::optional<c10::Device>) () from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cuda_cu.so
#10 0x00007fff836c3c77 in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (at::Tensor const&, c10::optional<c10::Device>), &at::(anonymous namespace)::(anonymous namespace)::wrapper___pin_memory>, at::Tensor, c10::guts::typelist::typelist<at::Tensor const&, c10::optional<c10::Device> > >, at::Tensor (at::Tensor const&, c10::optional<c10::Device>)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::Device>) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cuda_cu.so
#11 0x00007fffaf34f761 in at::_ops::_pin_memory::redispatch(c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::Device>) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#12 0x00007fffaf94958c in at::(anonymous namespace)::_pin_memory(at::Tensor const&, c10::optional<c10::Device>) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#13 0x00007fffaf949707 in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (at::Tensor const&, c10::optional<c10::Device>), &at::(anonymous namespace)::_pin_memory>, at::Tensor, c10::guts::typelist::typelist<at::Tensor const&, c10::optional<c10::Device> > >, at::Tensor (at::Tensor const&, c10::optional<c10::Device>)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::Device>) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#14 0x00007fffaf34f761 in at::_ops::_pin_memory::redispatch(c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::Device>) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#15 0x00007fffb0d6715e in torch::autograd::VariableType::(anonymous namespace)::_pin_memory(c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::Device>) () from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#16 0x00007fffb0d675aa in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::Device>), &torch::autograd::VariableType::(anonymous namespace)::_pin_memory>, at::Tensor, c10::guts::typelist::typelist<c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::Device> > >, at::Tensor (c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::Device>)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::Device>) () from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#17 0x00007fffaf3bdb1c in at::_ops::_pin_memory::call(at::Tensor const&, c10::optional<c10::Device>) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#18 0x00007fffaed5ddec in at::native::pin_memory(at::Tensor const&, c10::optional<c10::Device>) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#19 0x00007fffafcc0197 in c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor (at::Tensor const&, c10::optional<c10::Device>), &at::(anonymous namespace)::(anonymous namespace)::wrapper__pin_memory>, at::Tensor, c10::guts::typelist::typelist<at::Tensor const&, c10::optional<c10::Device> > >, at::Tensor (at::Tensor const&, c10::optional<c10::Device>)>::call(c10::OperatorKernel*, c10::DispatchKeySet, at::Tensor const&, c10::optional<c10::Device>) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#20 0x00007fffaf3bd78c in at::_ops::pin_memory::call(at::Tensor const&, c10::optional<c10::Device>) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_cpu.so
#21 0x00007fffd88d1f89 in torch::autograd::THPVariable_pin_memory(_object*, _object*, _object*) ()
   from /home/tjing/anaconda3/envs/fp8/lib/python3.8/site-packages/torch/lib/libtorch_python.so
#22 0x00005555556f91c7 in method_vectorcall_VARARGS_KEYWORDS (func=<optimized out>, args=<optimized out>, nargsf=<optimized out>,
    kwnames=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Objects/descrobject.c:332
#23 0x000055555565475e in _PyObject_Vectorcall (kwnames=0x0, nargsf=<optimized out>, args=0x7fff7507c3a8, callable=0x7ffff73ee680)
    at /tmp/build/80754af9/python_1599203911753/work/Include/cpython/abstract.h:127
#24 call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>, tstate=0x55555a523880)
    at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:4963
#25 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
    at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:3486
#26 0x00005555556def9f in _PyEval_EvalCodeWithName (_co=0x7fff2e6da190, globals=<optimized out>, locals=<optimized out>,
    args=<optimized out>, argcount=<optimized out>, kwnames=0x0, kwargs=0x7fff75076c68, kwcount=<optimized out>, kwstep=1,
    defs=0x7fff2e74c568, defcount=1, kwdefs=0x0, closure=0x0, name=0x7ffff77013f0, qualname=0x7ffff77013f0)
    at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:4298
#27 0x00005555556df943 in _PyFunction_Vectorcall (func=<optimized out>, stack=0x7fff75076c58, nargsf=<optimized out>,
    kwnames=<optimized out>) at /tmp/build/80754af9/python_1599203911753/work/Objects/call.c:435
#28 0x0000555555652b84 in _PyObject_Vectorcall (kwnames=0x0, nargsf=<optimized out>, args=0x7fff75076c58, callable=0x7fff2e6d9ca0)
    at /tmp/build/80754af9/python_1599203911753/work/Include/cpython/abstract.h:127
#29 call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>, tstate=0x55555a523880)
    at /tmp/build/80754af9/python_1599203911753/work/Python/ceval.c:4963
#30 _PyEval_EvalFrameDefault (f=<optimized out>, throwflag=<optimized out>)
```

Any idea what's wrong? Is an intel CPU required if I intend to run the model on GPU only?

I tried varying the batch size to no avail.

**Comment by tjingrant** [comment]
Actually it seems like the calibration step requires a very small batch size (e.g., 32) to work. Looking at nvidia-smi it barely uses more than 25% of my available GPU memory (~6GB/24GB). Any tip for making the calibration step more efficient?

**Comment by nkmellem** [comment]
This might be something do with how memory is managed by CUDA. During calibration, the model is switched to training model (model.train() ), cuDNN might be taking this signal to allocating large memory pools of memory to accommodate gradients.  
You can quickly test this by switching to model.test() before the calibration step.  Can you try that and report back?
